### PR TITLE
External Ad Serving

### PR DESCRIPTION
**Pull request by anila-niaz**
Replace click_urls with click_urls_info in ad renderer request.

https://github.com/MEDIARITHMICS/mediarithmics-platform/pull/49

### CODE REVIEW

**Comment by thomastosoni** [comment]
We should agree to wait until spec and all dev is done before making the pull requests. For example, I changed `escapes` in clickUrlsInfo to `encodingCount` because escaping != encoding, this isn't changed in the spec yet and now it's in here. Also since we can't merge without the other prs it adds clutter to the list of we want to cleanup. @anila-niaz @PoulpiFr

**Comment by PoulpiFr** [comment]
@thomastosoni > Hum, I'm not 100% sure about `encoding_count` for 2 reasons:

## 1. APIs alignement

One thing that we know is that we are walking on Google DCM territory here (DCM is the Google AdServer, a leader of the market). And they've been talking to all their users about `escaping`: [See `data-dcm-landing-page-escapes`](https://support.google.com/dcm/partner/answer/6080468?hl=fr#table).

So, I'm pretty sure that when we'll have the equivalent of their `<ins>` tag for the AdServer, we'll use the same terminology (easier to use the same words / concepts than Google here).

Then the question is about keeping or not an alignment between the AdRenderer developer API (the /ad_contents request payload) and the end-user API (the `ins` Tag parameters names). Personally, I'm in favor of staying aligned whenever possible.

 ## 2. AdRenderer dev comprehension

Lets take a simple example:
```
click_urls_infos: [
{
"url": "https://url1.com/click?redirect=",
"encoding_count": 1
},
{
"url": "https://url2.com",
"encoding_count": 0
}]
```

Ad an AdRenderer developer, I'm not sure if it is clear that `click_urls_info[0].encoding_count` is the number of times that I should encode `click_urls_info[1].url` before appending it to `click_urls_info[0].url`.

In my opinion, `encoding_count`:
1. is not asking me to encode something but is rather describing me how many time something was already encoded.
2. doesn't really explain that this value should be used for the next item in the array

So, here are sone other proposals:
`escapes_needed` / `encoding_level` / `redirect_count`

`redirect_count` is my favorite right now: if the AdRenderer dev know how are working URL redirect chain, this is telling him everything. Which could be easier that trying to explain him what to do with the value with `encoding` / `escapes` terms.

What do you think?

**Comment by thomastosoni** [comment]
This is a long explanation to counter `encoding_count`! But ok, I prefer `redirect_count` too. I changed it in the backend pr, @anila-niaz can you change it again please?

**Comment by PoulpiFr** [comment]
@thomastosoni > I was thinking while writing the explanation: I didn't have the `redirect_count` idea until I tried to write all the things I didn't like about `encoding_count` =)

@anila-niaz > LGTM!

Can you add a line in the `CHANGELOG.md` about the deprecated `click_urls` field and the introduction of `click_urls_info` instead?

Also, in the `README.md` can you add a small section to describe the migration to the `v0.6` of the PluginSDK for the AdRenderer plugins?
- `click_urls_info` instead of `click_urls`
- Change of signature of `AdRendererBasePlugin.getEncodedClickUrl()` (`string` to `ClickUrlsInfo`)

No need to talk about the HandlebarsEngine changes: they're not breaking any API from what I'm seeing. Only the internals have changed.

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 - Rename the `recommenderProperties` field to `properties` for Recommender support
 - Change the type of `properties` in the Instance Context of Audience Feed & Recommender from `PluginProperty[]` to `PropertiesWrapper`
+- Replace `click_url` field with `click_urls_info` in AdRendererRequest, which contains the property `redirect_count` in addition to `url` for each entry.
```
Comment (PoulpiFr) [inline CHANGELOG.md]: click_url**s**

File: README.md [hunk 2]
```diff
@@ -241,4 +241,34 @@ Prior to the v0.3.0, there was only one Handlebars engine provided in the extra
 
 With the 0.3.0+, there are now 2 Handlebars engine:
 - HandlebarsEngine: to be used when building an AdRenderer without recommendations
-- RecommendationsHandlebarsEngine: to be used when building a 'recommendation' Ad Renderer
\ No newline at end of file
+- RecommendationsHandlebarsEngine: to be used when building a 'recommendation' Ad Renderer
+
+## Migration from 0.5.x to 0.6.x
+* `click_url` property of `AdRendererRequest` is replaced with `click_urls_info`.
```
Comment (PoulpiFr) [inline README.md]: click_url**s**

**Pull request merged**